File: krono_core_frontend_movil/lib/data/api_client.dart
/*
pi_client.dart — Cliente HTTP principal (`ApiClient`) responsable de centralizar y gestionar todas las
comunicaciones con la API REST del backend. Maneja automáticamente los tiempos de espera,
la serialización de datos y la conversión de formatos, proveyendo métodos estáticos
estandarizados (`get`, `post`, `put`, `delete`) para que los utilicen el resto de servicios.

 RESPONSABILIDADES:
 - Resuelve la URL base del backend según la plataforma (web, desktop, Android)
 - Añade cabeceras JWT Bearer automáticamente
- Añade cabeceras de contexto multi-tenant (X-Empresa-Id, X-User-Rol)
- Auto-refresh del token cuando expira (detecta 401 y reintenta)
- Métodos HTTP: get(), post(), put(), patch(), delete()
*/

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../global/utils/session.dart';
import '../global/models/models.dart';

/// CLASE: ApiClient
/// Centraliza todas las peticiones HTTP de la aplicación.
///
/// URL base dinámica según la plataforma:
/// - Web, Desktop y móvil: apunta al backend de Azure.
/// - Si necesitas desarrollo local, reemplaza solo esta URL o usa tu propia configuración.
class ApiClient {
  static const String _remoteBaseUrl =
      'https://kronobackend-ctach7cya4cxb9d8.spaincentral-01.azurewebsites.net/api';

  /// URL base del backend usada por toda la app.
  static String get baseUrl => _remoteBaseUrl;

  /// Construye las cabeceras JSON con JWT Bearer + contexto multi-tenant.
  ///
  /// Cabeceras incluidas:
  /// - Content-Type y Accept: application/json
  /// - Authorization: Bearer `<access_token>` (JWT)
  /// - X-User-Rol: rol del usuario (soporte legacy)
  /// - X-Empresa-Id: ID de empresa (soporte legacy)
  static Map<String, String> _jsonHeaders() {
    final u = SessionController.instance.user;
    final h = <String, String>{
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    /// Token JWT de acceso
    final token = SessionController.instance.accessToken;
    if (token != null) {
      h['Authorization'] = 'Bearer $token';
    }

    /// Contexto multi-tenant (soporte legacy, JWT ya lleva rol y empresaId)
    if (u != null) {
      h['X-User-Rol'] = u.rol.name;
      if (u.empresaId != null &&
          u.rol != UserRole.cliente &&
          u.rol != UserRole.admin) {
        h['X-Empresa-Id'] = u.empresaId.toString();
      }
      if (u.rol == UserRole.admin &&
          SessionController.instance.adminEmpresaId != null) {
        h['X-Empresa-Id'] = SessionController.instance.adminEmpresaId
            .toString();
      }
    }
    return h;
  }

  /// Realiza una petición GET al endpoint especificado.
  /// Si recibe 401, intenta refrescar el token y reintenta automáticamente.
  static Future<http.Response> get(String endpoint) async {
    final url = Uri.parse('$baseUrl$endpoint');
    var response = await http.get(url, headers: _jsonHeaders());
    if (response.statusCode == 401 &&
        SessionController.instance.refreshToken != null) {
      await SessionController.instance.refreshAccessToken();
      response = await http.get(url, headers: _jsonHeaders());
    }
    return response;
  }

  /// Realiza una petición POST con datos JSON.
  static Future<http.Response> post(
    String endpoint,
    Map<String, dynamic> data,
  ) async {
    final url = Uri.parse('$baseUrl$endpoint');
    var response = await http.post(
      url,
      headers: _jsonHeaders(),
      body: jsonEncode(data),
    );
    if (response.statusCode == 401 &&
        SessionController.instance.refreshToken != null) {
      await SessionController.instance.refreshAccessToken();
      response = await http.post(
        url,
        headers: _jsonHeaders(),
        body: jsonEncode(data),
      );
    }
    return response;
  }

  /// Realiza una petición PUT con datos JSON (actualización completa).
  static Future<http.Response> put(
    String endpoint,
    Map<String, dynamic> data,
  ) async {
    final url = Uri.parse('$baseUrl$endpoint');
    var response = await http.put(
      url,
      headers: _jsonHeaders(),
      body: jsonEncode(data),
    );
    if (response.statusCode == 401 &&
        SessionController.instance.refreshToken != null) {
      await SessionController.instance.refreshAccessToken();
      response = await http.put(
        url,
        headers: _jsonHeaders(),
        body: jsonEncode(data),
      );
    }
    return response;
  }

  /// Realiza una petición PATCH con datos JSON (actualización parcial).
  static Future<http.Response> patch(
    String endpoint,
    Map<String, dynamic> data,
  ) async {
    final url = Uri.parse('$baseUrl$endpoint');
    var response = await http.patch(
      url,
      headers: _jsonHeaders(),
      body: jsonEncode(data),
    );
    if (response.statusCode == 401 &&
        SessionController.instance.refreshToken != null) {
      await SessionController.instance.refreshAccessToken();
      response = await http.patch(
        url,
        headers: _jsonHeaders(),
        body: jsonEncode(data),
      );
    }
    return response;
  }

  /// Realiza una petición DELETE al endpoint especificado.
  static Future<http.Response> delete(String endpoint) async {
    final url = Uri.parse('$baseUrl$endpoint');
    var response = await http.delete(url, headers: _jsonHeaders());
    if (response.statusCode == 401 &&
        SessionController.instance.refreshToken != null) {
      await SessionController.instance.refreshAccessToken();
      response = await http.delete(url, headers: _jsonHeaders());
    }
    return response;
  }
}

File: kronoCore_Frontend_Web/global/js/api_client.js
/*
 * api_client.js — Cliente HTTP centralizado para el API REST de KronoCore.
 * 
 * FUNCIONES PRINCIPALES:
 * - Gestiona la URL base de la API (https://kronobackend-ctach7cya4cxb9d8.spaincentral-01.azurewebsites.net/api)
 * - Añade cabeceras JWT Bearer automáticamente (Authorization)
 * - Añade cabeceras de contexto multi-tenant (X-Empresa-Id, X-User-Rol)
 * - Auto-refresh de token JWT cuando expira (detecta 401 y reintenta)
 * - Métodos HTTP: get(), post(), put(), patch(), delete()
 * 
 * USO: ApiClient.get('/clientes'), ApiClient.post('/clientes', data)
 */

/**
 * api_client.js
 * Cliente HTTP simple para consumir el backend Spring Boot con JWT.
 */

// Cabeceras legacy de tenant (mantenido para exportarCSV y compatibilidad)
function tenantHeaders() {
  const u = typeof SessionController !== 'undefined' ? SessionController.instance.user : null;
  if (!u) return {};
  const h = {};
  if (u.rol) h['X-User-Rol'] = u.rol;
  if (u.empresaId != null && u.rol !== 'cliente' && u.rol !== 'admin') h['X-Empresa-Id'] = String(u.empresaId);
  if (u.rol === 'admin' && SessionController.instance.adminEmpresaId != null) h['X-Empresa-Id'] = String(SessionController.instance.adminEmpresaId);
  return h;
}

// Cabeceras de autenticación JWT + contexto de empresa
function authHeaders() {
  const u = typeof SessionController !== 'undefined' ? SessionController.instance.user : null;
  if (!u) return {};
  const h = {};
  // Token JWT de acceso - se envía en cada petición como Bearer
  const token = SessionController.instance.accessToken;
  if (token) h['Authorization'] = 'Bearer ' + token;
  // Contexto multi-tenant (soporte legacy, JWT ya lleva rol y empresaId)
  if (u.rol) h['X-User-Rol'] = u.rol;
  if (u.empresaId != null && u.rol !== 'cliente' && u.rol !== 'admin') h['X-Empresa-Id'] = String(u.empresaId);
  if (u.rol === 'admin' && SessionController.instance.adminEmpresaId != null) h['X-Empresa-Id'] = String(SessionController.instance.adminEmpresaId);
  return h;
}

// Objeto con la configuración base y los métodos de petición al servidor
const ApiClient = {
  baseUrl: 'https://kronobackend-ctach7cya4cxb9d8.spaincentral-01.azurewebsites.net/api',

  /**
   * Petición genérica: concatena baseUrl + path, parsea JSON y propaga errores HTTP.
   * 
   * FLUJO:
   * 1. Realiza la petición HTTP con fetch()
   * 2. Parsea la respuesta como JSON (si hay contenido)
   * 3. Si la respuesta es 401 (token expirado), intenta refrescar automáticamente
   * 4. Si el refresh falla, redirige al login
   * 5. Si la respuesta es OK, devuelve los datos
   * 
   * @param {string} path - Ruta del endpoint (ej: '/clientes')
   * @param {object} options - Opciones de fetch (method, body, etc.)
   * @returns {Promise<object>} - Datos parseados de la respuesta
   */
  async request(path, options = {}) {
    const response = await fetch(`${this.baseUrl}${path}`, {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        ...authHeaders(),
        ...(options.headers || {}),
      },
      ...options,
    });

    let body = null;
    const text = await response.text();
    if (text) {
      try {
        body = JSON.parse(text);
      } catch {
        body = text;
      }
    }

    if (!response.ok) {
      const message = (body && body.message) || `HTTP ${response.status}`;
      // Si el token expiró (401), intentar refrescar automáticamente
      if (response.status === 401 && SessionController.instance.refreshToken) {
        try {
          await SessionController.instance.refreshAccessToken();
          // Reintentar la petición original con el nuevo token
          const retryHeaders = { ...authHeaders(), ...(options.headers || {}) };
          const retryResp = await fetch(`${this.baseUrl}${path}`, {
            headers: {
              'Content-Type': 'application/json',
              Accept: 'application/json',
              ...retryHeaders,
            },
            ...options,
          });
          const retryText = await retryResp.text();
          if (retryText) {
            try { return JSON.parse(retryText); } catch { return retryText; }
          }
          return null;
        } catch {
          // Refresh falló: redirigir a login
          SessionController.instance.logout();
          window.location.href = '../login/login.html';
          throw new Error('Sesión expirada');
        }
      }
      throw new Error(message);
    }

    return body;
  },

  /** GET: obtener datos del servidor */
  get(path) { return this.request(path, { method: 'GET' }); },
  
  /** POST: crear un nuevo recurso */
  post(path, data) { return this.request(path, { method: 'POST', body: JSON.stringify(data) }); },
  
  /** PUT: actualizar un recurso completo */
  put(path, data) { return this.request(path, { method: 'PUT', body: JSON.stringify(data) }); },
  
  /** PATCH: actualización parcial de un recurso */
  patch(path, data) { return this.request(path, { method: 'PATCH', body: JSON.stringify(data) }); },
  
  /** DELETE: eliminar un recurso */
  delete(path) { return this.request(path, { method: 'DELETE' }); },
};